### PR TITLE
developmentPullRequest

### PR DESCRIPTION
**Pull request by IvonneBenitesRodriguez**
**About this project**🧑🏽‍🍳 <br/>
The project is about a Recipe app which keeps track of all our user's recipes, ingredients, and inventory. It will allow our users to save ingredients, keep track of what they have, create recipes, and generate a shopping list based on what they have and what they are missing from a recipe. Also, since sharing recipes is an important part of cooking the app should allow our users to make them public so anyone can access them.<br/>

**For it:** ✅<br/>
We used Ruby gems as a software packages system.
We Installed Ruby on Rails framework.
We used controllers to handle requests and render empty views.
We used params from browser requests in a safe way.
We used preprocessed HTML file with embedded Ruby code.
We used layouts and templates for shared content.
We created database migration files to maintain database schema.
We built validations for models.
We used secure app from n+1 problems.
Between other functionalities added.
We passed the linters.

**Goal: 🧑🏽‍🍳**
The goal is to understand the usage of Rails, creating databases, and migrations files, and most of all feeling comfortable building an App like Recipe App.

### CODE REVIEW

File: app/views/layouts/_navbar.html.erb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<nav class="navbar navbar-expand-lg navbar-light bg-success text-center">
+  <a class="navbar-brand" href="#">Recipe App</a>
+  <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
+    <span class="navbar-toggler-icon"></span>
+  </button>
+  <div class="collapse navbar-collapse" id="navbarNav">
+    <ul class="navbar-nav">
+      <li class="nav-item active">
+        <a class="nav-link" href="/foods">Add Food</span></a>
```
Comment (Lucas-Masaba) [inline app/views/layouts/_navbar.html.erb]: - [x] As mentioned in the project requirements, you should follow the layout of the [wireframe](https://raw.githubusercontent.com/microverseinc/curriculum-rails/main/recipe-app/images/food_index.png?token=GHSAT0AAAAAACJEKCOTHHLVIGHV2T6IZZE2ZNLATMA). This button should be moved to the right.
image

File: app/views/recipes/public_recipes.html.erb
Comment (Lucas-Masaba) [inline app/views/recipes/public_recipes.html.erb:1]: - [ ] You are required to follow the layout of the [wireframe](https://raw.githubusercontent.com/microverseinc/curriculum-rails/main/recipe-app/images/public_recipe_index.png?token=GHSAT0AAAAAACJEKCOSGASE32D3ZRAS754OZNLBAMA) provided in the project description. Your public recipes are being displayed in a tabular format which is not depicted in the wireframe.
image

File: app/views/recipes/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<h1 class='text-center mt-5'><%= @recipe.name %></h1>
+
+<section class="text-center mt-5">
+  <%= render @recipe %>
+  <div class="container col-6 d-flex justify-content-between align-content-center my-5">
+    <%= button_to "Generate Shopping List", generate_shopping_list_recipe_path(@recipe), method: :post, class: "btn btn-primary" %>
```
Comment (Lucas-Masaba) [inline app/views/recipes/show.html.erb]: - [ ] When I try to generate a shopping list, I get nothing on the shopping list page.
image
image
Please make sure a user can generate a shopping list with the required information.

File: Gemfile
Comment (Lucas-Masaba) [inline Gemfile:1]: - [ ] You cannot use Rspec for testing if you do not have it present in your Gemfile. Please make sure to add it here.
image

**Review by Lucas-Masaba — CHANGES_REQUESTED** [review]
# Hello @IvonneBenitesRodriguez && @gilberthappi  👋 , 

Good job so far! 👍 
There are some issues that you still need to work on to go to the next project but you are almost there!

almost-there

## Highlights 🍾 
- Well documented README
- Nice UI :fireworks: 
- Met ost of the project requirements :gear: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions 💭 
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @Lucas-Masaba in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/recipes/public_recipes.html.erb
Comment (IvonneBenitesRodriguez) [reply app/views/recipes/public_recipes.html.erb:1]: Dear @Lucas-Masaba ,
Thank you for your suggestions, 
we modified it as you requested,
best regards,

File: app/views/recipes/show.html.erb (hunk 2 above)
Comment (IvonneBenitesRodriguez) [reply app/views/recipes/show.html.erb]: Dear @Lucas-Masaba ,
Thank you for your suggestions,
we modified it as you requested,
best regards,

File: Gemfile
Comment (IvonneBenitesRodriguez) [reply Gemfile:1]: Dear @Lucas-Masaba ,
Thank you for your suggestions,
we modified it as you requested,
best regards,

File: app/views/recipes/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        <td class="d-flex">
+          <%= link_to "Modify", edit_recipe_recipe_food_path(@recipe.id, ingredient.id), class: "btn border-danger text-danger" %>
+          <%= button_to "Delete", recipe_recipe_food_path(@recipe.id, ingredient.id) , method: :delete, class: "ms-2 btn border-danger text-danger" %>
+        </td>
+        <% end %>
+       </tr>
+      <% end %>
+    </tbody>
+  </table>
+</div>
```
Comment (Nemwel-Boniface) [inline app/views/recipes/show.html.erb]: - [ ] Goof job making the changes that were requested by the previous review. However, it seems that while making this requested change you removed the button for **"Generate shopping list"**. KIndly ensure that it is there as this is part of the projectrequirements.

**See below the provided wireframe:**

image

**This is how I see your project at the moment:**

image

File: Gemfile [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+# Use Active Storage variants [https://guides.rubyonrails.org/active_storage_overview.html#transforming-images]
+# gem "image_processing", "~> 1.2"
+
+group :development, :test do
+  # See https://guides.rubyonrails.org/debugging_rails_applications.html#debugging-with-the-debug-gem
+  gem 'database_cleaner'
+  gem 'debug', platforms: %i[mri mingw x64_mingw]
+  gem 'rails-controller-testing'
+  gem 'rspec-rails', '~> 6.0.0.rc1'
```
Comment (Nemwel-Boniface) [inline Gemfile:30]: - [ ] You did super well adding the gem rspec rails. However, when I run your written specs, I notice that there are some that are failing. Below is a screenshot of the run specs:

image

**The specific files that are facing this failure include:**
1. spec/views/food_index_spec.rb
2. spec/views/public_recipes_spec.rb
3. spec/views/recipes_index_spec.rb
4. spec/views/recipes_show_spec.rb
5. spec/views/shopping_list_spec.rb

**KIndly ensure that you resolve all pending requirements before resubmitting for review.**

**Review by Nemwel-Boniface — CHANGES_REQUESTED** [review]
## Hello @IvonneBenitesRodriguez  and @gilberthappi    :wave: , 
I'm Nemwel your reviewer for this PR :man_beard: 

Great job addressing some of the changes that were requested by a previous reviewer :clap: 
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights

Great job setting up the correct linters ✔️ 
Professional documentation ✔️ 
Correct workflow ✔️ 
Meaningful commit messages ✔️  


## ♻️ Required Changes ♻️

_Check the comments under the review._

## [Optional] suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding! :partying_face: 👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me **@Nemwel-Boniface** in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: Gemfile (hunk 4 above)
Comment (IvonneBenitesRodriguez) [reply Gemfile:30]: Thank you @Nemwel-Boniface very nice to meet you,
We modified it as you requested, 
Bests,

File: app/views/recipes/show.html.erb (hunk 3 above)
Comment (IvonneBenitesRodriguez) [reply app/views/recipes/show.html.erb]: Thank you @Nemwel-Boniface very nice to meet you,
We modified it as you requested,
Bests,

**Review by Munsa1 — APPROVED** [review]
Hi @IvonneBenitesRodriguez @gilberthappi , 
congrats group capstone

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 


Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Comment by gilberthappi** [comment]
Thanks alot

**Pull request merged**